File: framework/languages/node/e2e/PubSub/Client/Scenarios/ps-a3-late-subscriber-scenario.ts
// PS-A3: Late subscriber는 ready 이후 event부터 받는다 시나리오를 검증한다.
import { randomUUID } from 'node:crypto';
import { PubSubNames } from '../../Shared/messages';
import { getJson } from '../../../http-client';
import { NetworkFaultProxy } from '../Support/network-fault-proxy';
import { ensure } from '../Support/scenario-assert';
import { ServerProcessLauncher } from '../Support/server-process-launcher';
import { publishEvent, publishUntilDelivered } from './ps-a1-fanout-basic-delivery-scenario';

export async function runPsA3(
  publisher: string,
  lateSubscriberUrl: string,
  processes: ServerProcessLauncher,
  publisherEndpoint: string
): Promise<void> {
  const beforeReadyRun = randomUUID().replaceAll('-', '');
  await publishEvent(publisher, PubSubNames.mainTopic, beforeReadyRun, 1, 'before-ready');

  const fault = await NetworkFaultProxy.start(publisherEndpoint, true);
  const subscriber = processes.startSubscriber(
    'sub-late', lateSubscriberUrl, 'sub-late.evidence.log', fault.endpoint
  );
  try {
    await subscriber.waitReady();
    fault.unblock();
    const afterReadyRun = await publishUntilDelivered(
      publisher, lateSubscriberUrl, PubSubNames.mainTopic, 2, 'after-ready'
    );
    const evidence = await getJson<readonly string[]>(lateSubscriberUrl, '/evidence');
    ensure(
      evidence.some((line) => line.includes(`run=${afterReadyRun}`))
        && evidence.every((line) => !line.includes(`run=${beforeReadyRun}`)),
      'PS-A3 late subscriber replayed before-ready or missed the first after-ready event.'
    );
    console.log('scenario PS-A3 passed');
  } finally {
    await subscriber.kill();
    await fault.close();
  }
}

File: framework/languages/dotnet/e2e/PubSub/Client/Scenarios/PsA3LateSubscriberScenario.cs
// Verifies PS-A3 Late Subscriber behavior.
using PubSub.Client.Support;
using PubSub.Shared;
using Zlink.HttpClient;

namespace PubSub.Client.Scenarios;

// PS-A3: verifies that a subscriber started after publishing receives only post-join events.
internal static class PsA3LateSubscriberScenario
{
    public static async Task RunAsync(
        ZLinkHttpClient publisher,
        ZLinkHttpClient lateSubscriberClient,
        IReadOnlyList<ZLinkHttpClient> existingSubscribers,
        ServerProcessLauncher processes,
        string lateSubscriberUrl)
    {
        var beforeLateRun = Guid.NewGuid().ToString("N");

        // Publish before the late subscriber exists; this range must not be replayed later.
        for (var i = 1; i <= 5; i++)
            await publisher.Post("/publish/event")
                .Query("topic", PubSubNames.MainTopic)
                .Query("runId", beforeLateRun)
                .Query("sequence", i.ToString())
                .Query("value", $"before-late-{i}")
                .AsyncRaw();

        // Keep the publisher transport blocked until the late subscriber host is
        // healthy, so the join it is measured against happens after this point.
        await using var connectionGate = new NetworkFaultProxy(
            new Uri(processes.PublisherEndpoint),
            initiallyBlocked: true);
        using var lateSubscriber = processes.StartSubscriber(
            "sub-late",
            lateSubscriberUrl,
            "sub-late.evidence.log",
            connectionGate.Endpoint.GetComponents(
                UriComponents.SchemeAndServer,
                UriFormat.Unescaped));
        try
        {
            await StateObservation.WaitUntilAsync(
                async () =>
                {
                    try
                    {
                        return (await lateSubscriberClient.Get("/health").AsyncRaw()).Status == 200;
                    }
                    catch (Exception ex) when ((ex is HttpRequestException || ex.InnerException is HttpRequestException))
                    {
                        return false;
                    }
                },
                "PS-A3 expected late subscriber to become healthy.");
            connectionGate.Unblock();
            await connectionGate.WaitForUpstreamConnectionAsync();
            await FanoutReadiness.WaitUntilReceivingAsync(
                publisher, lateSubscriberClient, "PS-A3 late subscriber");

            var afterLateRun = Guid.NewGuid().ToString("N");

            // Probe delivery proves the late subscriber joined before this measured publish.
            await publisher.Post("/publish/event")
                .Query("topic", PubSubNames.MainTopic)
                .Query("runId", afterLateRun)
                .Query("sequence", "1")
                .Query("value", "after-late")
                .AsyncRaw();

            var measuredSubscribers = existingSubscribers.Append(lateSubscriberClient).ToArray();
            var measuredEvidence = await Task.WhenAll(measuredSubscribers.Select(
                subscriber => SubscriberObservation.WaitForEventAsync(subscriber, afterLateRun, 1)));
            var lateEvidence = measuredEvidence[^1];

            // Pub/Sub has no replay contract, so pre-join events must stay absent from late evidence.
            ZlinkStreamAssert.Ensure(
                lateEvidence.All(line => !line.Contains($"run={beforeLateRun}", StringComparison.Ordinal)),
                "PS-A3 late subscriber replayed pre-join events.");
            Console.WriteLine("scenario PS-A3 passed");
        }
        finally
        {
            if (!lateSubscriber.HasExited)
            {
                lateSubscriber.Kill(true);
                await lateSubscriber.WaitForExitAsync();
            }
        }
    }
}
